tests/test_twitter_ads_bookmark.py

```diff
@@ -1,179 +1,218 @@
+                    # We are writing a separate bookmark for the child stream in which we are storing
+                    # the bookmark based on the parent's replication key.
+                    # But, we are not using any fields from the child record for it.
+                    # In addition to it, we are just using parent's primary key to fetch the child record but
+                    # we are not storing parent record any where.
+                    if stream == "targeting_criteria":
+                        # collect parent pk value for 1st sync
+                        first_sync_child_foreign_keys = set()
+                        for record in first_sync_messages:
+                            first_sync_child_foreign_keys.add(record['line_item_id'])
+
+                        # collect parent pk value for 2nd sync
+                        second_sync_child_foreign_keys = set()
+                        for record in second_sync_messages:
+                            second_sync_child_foreign_keys.add(record['line_item_id'])
+
+                        # Verify that all foreign keys in the 2nd sync are available in in 1st sync also. 
+                        self.assertTrue(second_sync_child_foreign_keys.issubset(first_sync_child_foreign_keys))
+
+                        # Collect child key-properties from 1st sync records based on foreign keys of 2nd sync records
+                        first_sync_pks = set()
+                        for record in first_sync_messages:
+                            if record['line_item_id'] in second_sync_child_foreign_keys:
+                                first_sync_pks.add((record['id'], record['line_item_id']))
+
+                        # Collect child key-properties from 2nd sync records
+                        second_sync_pks = set()
+                        for record in second_sync_messages:
+                            second_sync_pks.add((record['id'], record['line_item_id']))
+
+                        # Verify that each child record from 1st sync whose parent record has a replication key value
+                        # greater than or equal to bookmark value are replicated in 2nd sync.
+                        self.assertSetEqual(first_sync_pks, second_sync_pks)
+
+                        parent_sync_messages = [record.get('data') for record in
+                                                first_sync_records.get(
+                                                    "line_items", {}).get('messages', [])
+                                                if record.get('action') == 'upsert']
+
+                        # Verify all parents of the children in 2nd sync follows bookmark
+                        for record in parent_sync_messages:
+                            if record['id'] in second_sync_child_foreign_keys:
+                                self.assertGreaterEqual(record['updated_at'], second_bookmark_value)
```

Comment:
I think the logic here is incomplete. This assertion is saying "if a child is synced, the parent's replication key value must be greater/equal to the bookmark" which is good. But another assertion needs to be made that  "ANY parent record with a replication key value greater/equal to the bookmark MUST have it's child record synced". I think currently this could miss the case where the tap fails to replicate all  the child records that it should.

Reply:
We have added assertion to verify that all the child records of any parent with a replication key value greater/equal to the bookmark synced. All the parent records can not have child records. That's why we can not directly assert that all the child records of 1st sync parent are equal to 1st sync child record and the same thing for 2nd sync also.